# This notebook uses linux commands. Please run the cells using a linux system.

# Problem 1 - Word counts
## Part A. Characters in Little Women
### How many times are each of the following characters mentioned by name in the text of Little Women?

In [9]:
!wget https://raw.githubusercontent.com/gwsb-istm-6212-fall-2016/syllabus-and-schedule/master/projects/project-01/women.txt

--2016-09-21 12:45:32--  https://raw.githubusercontent.com/gwsb-istm-6212-fall-2016/syllabus-and-schedule/master/projects/project-01/women.txt
Resolving raw.githubusercontent.com... 151.101.32.133
Connecting to raw.githubusercontent.com|151.101.32.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1053440 (1.0M) [text/plain]
Saving to: ‘women.txt.3’

women.txt.3         100%[===================>]   1.00M   673KB/s    in 1.5s    

2016-09-21 12:45:34 (673 KB/s) - ‘women.txt.3’ saved [1053440/1053440]



In [10]:
!grep -oiE '\bJo\b|\bBeth\b|\bMeg\b|\bAmy\b' women.txt | sort | uniq -c

   7 AMY
 645 Amy
   8 BETH
 459 Beth
   7 JO
1355 Jo
   3 MEG
 683 Meg


'\b' in the above regular expression indicate word boundary. '-i' opiton for grep make it case insensitive. The counts are 645, 459, 1355, and 683 for Amy, Beth, Jo and Meg, respectively. But they do not include the counts for uppercase names, such as AMY.

## Part B. Juliet and Romeo in Romeo and Juliet
### How many times do each of the characters Juliet and Romeo have speaking lines in Romeo and Juliet?

In [11]:
!wget https://raw.githubusercontent.com/gwsb-istm-6212-fall-2016/syllabus-and-schedule/master/projects/project-01/romeo.txt

--2016-09-21 12:45:37--  https://raw.githubusercontent.com/gwsb-istm-6212-fall-2016/syllabus-and-schedule/master/projects/project-01/romeo.txt
Resolving raw.githubusercontent.com... 151.101.32.133
Connecting to raw.githubusercontent.com|151.101.32.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 178983 (175K) [text/plain]
Saving to: ‘romeo.txt.2’

romeo.txt.2         100%[===================>] 174.79K  --.-KB/s    in 0.1s    

2016-09-21 12:45:37 (1.57 MB/s) - ‘romeo.txt.2’ saved [178983/178983]



In [12]:
!grep -oE "Jul\.|Rom\." romeo.txt | sort | uniq -c

 117 Jul.
 163 Rom.


"Jul." and  "Rom." indicate speaking lines.
There are 117 and 163 speaking lines for Juliet and Romeo, respectively.

# Problem 2 - Capital Bikeshare
## Part A
### Which 10 Capital Bikeshare stations were the most popular departing stations in Q1 2016?

In [15]:
!wget https://raw.githubusercontent.com/gwsb-istm-6212-fall-2016/syllabus-and-schedule/master/projects/project-01/2016q1.csv.zip

--2016-09-21 12:46:05--  https://raw.githubusercontent.com/gwsb-istm-6212-fall-2016/syllabus-and-schedule/master/projects/project-01/2016q1.csv.zip
Resolving raw.githubusercontent.com... 151.101.32.133
Connecting to raw.githubusercontent.com|151.101.32.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 10643003 (10M) [application/octet-stream]
Saving to: ‘2016q1.csv.zip’

2016q1.csv.zip      100%[===================>]  10.15M  1.63MB/s    in 7.5s    

2016-09-21 12:46:13 (1.35 MB/s) - ‘2016q1.csv.zip’ saved [10643003/10643003]



### "-o" option for unzip overwrites the files if exists. 

In [16]:
!unzip -o 2016q1.csv.zip

Archive:  2016q1.csv.zip
  inflating: 2016q1.csv              


In [17]:
!csvcut -n 2016q1.csv

  1: Duration (ms)
  2: Start date
  3: End date
  4: Start station number
  5: Start station
  6: End station number
  7: End station
  8: Bike number
  9: Member Type


In [18]:
!csvcut 2016q1.csv | head -5 | csvlook

|----------------+-----------------+---------------+----------------------+--------------------------------+--------------------+---------------------------+-------------+--------------|
|  Duration (ms) | Start date      | End date      | Start station number | Start station                  | End station number | End station               | Bike number | Member Type  |
|----------------+-----------------+---------------+----------------------+--------------------------------+--------------------+---------------------------+-------------+--------------|
|  301295        | 3/31/2016 23:59 | 4/1/2016 0:04 | 31280                | 11th & S St NW                 | 31506              | 1st & Rhode Island Ave NW | W00022      | Registered   |
|  557887        | 3/31/2016 23:59 | 4/1/2016 0:08 | 31275                | New Hampshire Ave & 24th St NW | 31114              | 18th St & Wyoming Ave NW  | W01294      | Registered   |
|  555944        | 3/31/2016 23:59 | 4/1/2016 0:08 | 31101       

The below commands select "start station", the fifth column, sort the values of the column before using "uniq -c" to count unique values, then sort the unique values by count in a reverse order and print out the top 10 values.

In [19]:
!csvcut -c5 2016q1.csv | sort | uniq -c | sort -rn | head -10

13120 Columbus Circle / Union Station
9560 Massachusetts Ave & Dupont Circle NW
9388 Lincoln Memorial
8138 Jefferson Dr & 14th St SW
7479 Thomas Circle
7401 15th & P St NW
6568 14th & V St NW
6491 New Hampshire Ave & T St NW
5649 Eastern Market Metro / Pennsylvania Ave & 7th St SE
5514 17th & Corcoran St NW


###  Which 10 were the most popular destination stations in Q1 2016?

Following the same logic for starting station

In [20]:
!csvcut -c7 2016q1.csv | sort | uniq -c | sort -rn | head -10

13880 Columbus Circle / Union Station
11183 Massachusetts Ave & Dupont Circle NW
9419 Lincoln Memorial
8975 Jefferson Dr & 14th St SW
8092 15th & P St NW
7267 14th & V St NW
6997 Thomas Circle
6245 New Hampshire Ave & T St NW
5761 5th & K St NW
5651 17th & Corcoran St NW


## Part B
### For the most popular departure station, which 10 bikes were used most in trips departing from there?

In [21]:
#filter the starting station for "Columbus Circle / Union Station", then find out which 10 bikes were used most.
!csvgrep -c5 -m "Columbus Circle / Union Station" 2016q1.csv | csvcut -c8 | sort | uniq -c | sort -rn | head -10

  17 W22227
  16 W21867
  16 W21641
  16 W21538
  16 W21239
  16 W20540
  16 W00714
  15 W22080
  15 W21450
  15 W21076


### Which 10 bikes were used most in trips ending at the most popular destination station?

In [22]:
!csvgrep -c7 -m "Columbus Circle / Union Station" 2016q1.csv | csvcut -c8 | sort | uniq -c | sort -rn | head -10

  18 W00485
  17 W22227
  16 W22099
  16 W22080
  16 W21239
  16 W21076
  16 W20425
  16 W00714
  15 W21997
  15 W21867


# Problem 3 - Filters
## Part A
## ! The .py files are written in Python 3 syntax

In [23]:
!grep -oE '\w{{2,}}' women.txt | tr '[:upper:]' '[:lower:]' | sort | uniq -c | sort -rn | head -10

8155 and
7689 the
5152 to
3523 of
3245 her
2774 it
2503 in
2447 you
2343 she
2233 for
sort: write failed: standard output: Broken pipe
sort: write error


In [24]:
#save user python path
%sc my_python=which python


In [25]:
%%writefile grep_word.py

import fileinput
import re
"""
A word splitter 
"""
for line in fileinput.input():
    for word in re.findall(r'\w{2,}', line):
        print(word)


Overwriting grep_word.py


In [26]:
#add shebang line
!echo "#!{my_python}`cat grep_word.py`" > grep_word.py
!chmod +x grep_word.py

In [27]:
%%writefile lower.py

import fileinput

"""
Convert words into lowercase
"""
for line in fileinput.input():
        #remove line break
        line = line[:-1]
        print(line.lower())



Overwriting lower.py


In [28]:
#add shebang line
!echo "#!{my_python}`cat lower.py`" > lower.py
!chmod +x lower.py

In [29]:
!./grep_word.py women.txt | ./lower.py | sort | uniq -c | sort -rn | head -10

8155 and
7689 the
5152 to
3523 of
3245 her
2774 it
2503 in
2447 you
2343 she
2233 for
sort: write failed: standard output: Broken pipe
sort: write error


## Part B

In [30]:
#Download a stop-word list
!wget http://xpo6.com/wp-content/uploads/2015/01/stop-word-list.csv

--2016-09-21 12:48:33--  http://xpo6.com/wp-content/uploads/2015/01/stop-word-list.csv
Resolving xpo6.com... 104.28.21.102, 104.28.20.102, 2400:cb00:2048:1::681c:1466, ...
Connecting to xpo6.com|104.28.21.102|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 669 [text/csv]
Saving to: ‘stop-word-list.csv.1’

stop-word-list.csv. 100%[===================>]     669  --.-KB/s    in 0s      

2016-09-21 12:48:33 (37.5 MB/s) - ‘stop-word-list.csv.1’ saved [669/669]



In [31]:
#read the csv to a list then print out
with open("stop-word-list.csv",'r') as f:
    #it has only one line
    stop_words = [word.strip() for word in f.read().strip().split(',')]
    print(stop_words)

['a', 'able', 'about', 'across', 'after', 'all', 'almost', 'also', 'am', 'among', 'an', 'and', 'any', 'are', 'as', 'at', 'be', 'because', 'been', 'but', 'by', 'can', 'cannot', 'could', 'dear', 'did', 'do', 'does', 'either', 'else', 'ever', 'every', 'for', 'from', 'get', 'got', 'had', 'has', 'have', 'he', 'her', 'hers', 'him', 'his', 'how', 'however', 'i', 'if', 'in', 'into', 'is', 'it', 'its', 'just', 'least', 'let', 'like', 'likely', 'may', 'me', 'might', 'most', 'must', 'my', 'neither', 'no', 'nor', 'not', 'of', 'off', 'often', 'on', 'only', 'or', 'other', 'our', 'own', 'rather', 'said', 'say', 'says', 'she', 'should', 'since', 'so', 'some', 'than', 'that', 'the', 'their', 'them', 'then', 'there', 'these', 'they', 'this', 'tis', 'to', 'too', 'twas', 'us', 'wants', 'was', 'we', 'were', 'what', 'when', 'where', 'which', 'while', 'who', 'whom', 'why', 'will', 'with', 'would', 'yet', 'you', 'your']


Copy it into the below filter function so what we do not read the csv everytime.

In [32]:
%%writefile remove_stop_word.py

import fileinput

"""
A Filter For Removing Stop Words
"""

STOP_WORDS = ['a', 'able', 'about', 'across', 'after', 'all', 'almost', 'also', 'am', 'among', 'an', 'and', 'any', 'are', 'as', 'at', 'be', 'because', 'been', 'but', 'by', 'can', 'cannot', 'could', 'dear', 'did', 'do', 'does', 'either', 'else', 'ever', 'every', 'for', 'from', 'get', 'got', 'had', 'has', 'have', 'he', 'her', 'hers', 'him', 'his', 'how', 'however', 'i', 'if', 'in', 'into', 'is', 'it', 'its', 'just', 'least', 'let', 'like', 'likely', 'may', 'me', 'might', 'most', 'must', 'my', 'neither', 'no', 'nor', 'not', 'of', 'off', 'often', 'on', 'only', 'or', 'other', 'our', 'own', 'rather', 'said', 'say', 'says', 'she', 'should', 'since', 'so', 'some', 'than', 'that', 'the', 'their', 'them', 'then', 'there', 'these', 'they', 'this', 'tis', 'to', 'too', 'twas', 'us', 'wants', 'was', 'we', 'were', 'what', 'when', 'where', 'which', 'while', 'who', 'whom', 'why', 'will', 'with', 'would', 'yet', 'you', 'your']

for word in fileinput.input():
    w = word.strip()
    if w not in STOP_WORDS:
        print(w)

    

Overwriting remove_stop_word.py


In [33]:
#add shebang line
!echo "#!{my_python}`cat remove_stop_word.py`" > remove_stop_word.py
!chmod +x remove_stop_word.py

In [34]:
!./grep_word.py women.txt | ./lower.py | ./remove_stop_word.py | sort | uniq -c | sort -rn | head -25

1362 jo
 730 little
 725 one
 686 meg
 658 up
 652 amy
 598 laurie
 550 don
 492 very
 485 out
 467 beth
 462 good
 399 now
 393 go
 380 old
 377 mother
 375 never
 374 much
 368 well
 361 ll
 360 see
 355 over
 346 more
 331 away
 321 time
sort: write failed: standard output: Broken pipe
sort: write error


# Extra Credit

In [37]:
!wget https://raw.githubusercontent.com/gwsb-istm-6212-fall-2016/syllabus-and-schedule/master/projects/project-01/texts.zip

--2016-09-21 12:49:21--  https://raw.githubusercontent.com/gwsb-istm-6212-fall-2016/syllabus-and-schedule/master/projects/project-01/texts.zip
Resolving raw.githubusercontent.com... 151.101.32.133
Connecting to raw.githubusercontent.com|151.101.32.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 12668137 (12M) [application/octet-stream]
Saving to: ‘texts.zip’

texts.zip           100%[===================>]  12.08M  1.76MB/s    in 8.4s    

2016-09-21 12:49:29 (1.44 MB/s) - ‘texts.zip’ saved [12668137/12668137]



In [38]:
!unzip -o texts.zip -d ./text

Archive:  texts.zip
  inflating: ./text/10001.txt        
  inflating: ./text/10002.txt        
  inflating: ./text/10003.txt        
  inflating: ./text/10004.txt        
  inflating: ./text/10005.txt        
  inflating: ./text/10006.txt        
  inflating: ./text/10007.txt        
  inflating: ./text/10008.txt        
  inflating: ./text/10009.txt        
  inflating: ./text/10010.txt        
  inflating: ./text/10011.txt        
  inflating: ./text/10012.txt        
  inflating: ./text/10013.txt        
  inflating: ./text/10014.txt        
  inflating: ./text/10015.txt        
  inflating: ./text/10016.txt        
  inflating: ./text/10017.txt        
  inflating: ./text/10018.txt        
  inflating: ./text/10019.txt        
  inflating: ./text/10020.txt        
  inflating: ./text/10021.txt        
  inflating: ./text/10023.txt        
  inflating: ./text/10024.txt        
  inflating: ./text/10025.txt        
  inflating: ./text/10026.txt        
  inflating: ./text/10027.txt 

### Remove all_words.txt if exists because the below parallel job uses '>>' to append to all_words.txt. We can get  wrong numbers if we run this notebook twice.  If all_words.txt does not exists,   '!rm all_words.txt' can generate an error, however, we can use '2>/dev/null', which redirects the error message to null, ignore this error. 

In [39]:
!rm all_words.txt 2>/dev/null 

In [40]:
!time ls ./text/*.txt \
    | parallel --eta -j+0 "grep -oE '\w{{2,}}' {{}} | tr '[:upper:]' '[:lower:]' | ./remove_stop_word.py >> all_words.txt"


Academic tradition requires you to cite works you base your article on.
When using programs that use GNU Parallel to process data for publication
please cite:

  O. Tange (2011): GNU Parallel - The Command-Line Power Tool,
  ;login: The USENIX Magazine, February 2011:42-47.

This helps funding further development; AND IT WON'T COST YOU A CENT.
If you pay 10000 EUR you should feel free to use GNU Parallel without citing.

To silence the citation notice: run 'parallel --citation'.


Computers / CPU cores / Max jobs to run
1:local / 8 / 8

Computer:jobs running/jobs completed/%of started jobs/Average seconds to complete
ETA: 0s Left: 0 AVG: 0.07s  local:0/108/100%/0.1s 

real	0m8.215s
user	0m55.024s
sys	0m2.897s


In [41]:
!time ./grep_word.py all_words.txt | sort | uniq -c | sort -rn | head -25

19587 one
12574 out
12233 up
10907 more
9685 man
9511 very
9502 now
8720 little
8606 time
8540 upon
7765 gutenberg
7489 see
7323 well
7264 project
7102 over
7018 such
7015 two
6891 down
6822 great
6781 old
6749 day
6670 work
6670 know
6566 good
6446 made
sort: write failed: standard output: Broken pipe
sort: write error

real	0m56.100s
user	0m58.243s
sys	0m0.190s
